src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,802 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+      compressed_min.push_back(unused_bits_min);
+      for(size_t i = 0; i < static_cast<uint8_t>(version_octets - discarded_octets_min); i++) {
+         compressed_min.push_back(min[i]);
+      }
+
+      compressed_max.push_back(unused_bits_max);
+      for(size_t i = 0; i < static_cast<uint8_t>(version_octets - discarded_octets_max); i++) {
+         compressed_max.push_back(max[i]);
+      }
+
+      into.start_sequence()
+         .add_object(ASN1_Type::BitString, ASN1_Class::Universal, compressed_min)
+         .add_object(ASN1_Type::BitString, ASN1_Class::Universal, compressed_max)
+         .end_cons();
+   }
+}
+
+template <IPAddressBlocks::Version V>
+void IPAddressBlocks::IPAddressOrRange<V>::decode_from(Botan::BER_Decoder& from) {
+   const size_t version_octets = static_cast<size_t>(V);
+
+   const ASN1_Type next_tag = from.peek_next_object().type_tag();
+
+   // this can either be a prefix or a single address
+   if(next_tag == ASN1_Type::BitString) {
+      // construct a min and a max address from the prefix
+
+      std::vector<uint8_t> prefix;
+      from.decode(prefix, ASN1_Type::OctetString, ASN1_Type::BitString, ASN1_Class::Universal);
+
+      // we have to account for the octet at the beginning that specifies how many bits are unused in the last octet
+      if(prefix.empty() || prefix.size() > version_octets + 1) {
+         throw Decoding_Error(
+            fmt("IP address range entries must have a length between 2 and {} bits.", static_cast<uint8_t>(V)));
+      }
+
+      const uint8_t unused = prefix.front();
+      const uint8_t discarded_octets = version_octets - (static_cast<uint8_t>(prefix.size()) - 1);
+
+      prefix.erase(prefix.begin());
+
+      for(size_t i = 0; i < discarded_octets; i++) {
+         prefix.push_back(0);
+      }
+
+      m_min = IPAddress<V>(prefix);
+
+      for(size_t i = version_octets - discarded_octets; i < version_octets; i++) {
+         prefix[i] = 0xff;
+      }
+
+      // set all unused bits to 1
+      for(size_t i = 0; i < unused; i++) {
+         prefix[version_octets - 1 - discarded_octets] |= (1 << i);
+      }
+
+      m_max = IPAddress<V>(prefix);
+   } else if(next_tag == ASN1_Type::Sequence) {
+      // this is a range
+
+      std::vector<uint8_t> addr_min;
+      std::vector<uint8_t> addr_max;
+
+      from.start_sequence()
+         .decode(addr_min, ASN1_Type::OctetString, ASN1_Type::BitString, ASN1_Class::Universal)
+         .decode(addr_max, ASN1_Type::OctetString, ASN1_Type::BitString, ASN1_Class::Universal)
+         .end_cons();
+
+      // address 1 (min)
+
+      // again accounting for the 'unused' octet at the beginning
+      if(addr_min.empty() || addr_min.size() > version_octets + 1) {
+         throw Decoding_Error(
+            fmt("IP address range entries must have a length between 1 and {} bytes.", static_cast<uint8_t>(V)));
+      }
+
+      addr_min.erase(addr_min.begin());  // remove 'unused' octet
```

Comment:
Further down for the max address we check that if `max_addr.empty()` then `unused` must be 0. We need to check this because we need to set the unused bits to 1. For the min address however they are already 0, so until now we've just ignored the unused octet, as we don't need to examine it. Should we still check it? In the case that `unused != 1` and `addr_min.empty()` it would technically be a "bad" encoding - I'm just not sure if the user should be made aware of that.

Reply:
Thanks for calling this out. My preference would be that decoding be as strict as possible unless there was a significant mis-encoding in the ecosystem, such that not supporting that mis-encoding would impair interop (as for example in #4732)